lib/utils/ignore.ts

```diff
@@ -295,11 +296,44 @@ export async function filterFilesWithDockerignore(
 	projectDir: string,
 	serviceDirsByService?: ServiceDirs,
 ): Promise<{ filteredFileList: FileStats[]; dockerignoreFiles: FileStats[] }> {
+	const { ignoreByDir, serviceDirs } = await getDockerignoreByDir(
+		projectDir,
+		serviceDirsByService,
+	);
+	const files = await listFiles(projectDir);
+	const dockerignoreFiles: FileStats[] = [];
+	const filteredFileList = files.filter((file: FileStats) => {
+		if (path.basename(file.relPath) === '.dockerignore') {
+			dockerignoreFiles.push(file);
+		}
+		for (const dir of serviceDirs) {
+			if (file.relPath.startsWith(dir)) {
+				return !ignoreByDir[dir].ignores(file.relPath.substring(dir.length));
+			}
+		}
+		return !ignoreByDir['.'].ignores(file.relPath);
+	});
+	return { filteredFileList, dockerignoreFiles };
+}
+
+/**
+ * Get dockerignore instances for root and all service directories in the project.
+ * Also return the list of service directories found.
+ * @param projectDir Source directory to
+ * @param serviceDirsByService Map of service names to their subdirectories.
+ */
+export async function getDockerignoreByDir(
```

Comment:
I was thinking about this function -- its input and output -- and I think it could be made more programmer friendly. Specially the return values: `ingoreByDir` and `serviceDirs`: how is one supposed to use these? The javadoc itself would be complex, to cover directory format requirements. One can figure it all out by carefully reading source code in multiple places, but it could be simpler. It made sense when these data structures were variable in the body of a function, but as an exported function for use in different files, I thought it could be clearer to index by service name rather than by directory. For one reason, service names are widely used and do not vary: a service name like "foo" or "main" is all there is to it, whereas a service directory may be represented in multiple ways, e.g.: `foo`, `./foo`, `./foo/`, `C:\Users\paulo\balena\other\simple_multicontainer\foo'`...

With this in mind, I will be shortly contributing a commit where I propose some changes.